### PR TITLE
Refactored texture compiler to use JNI

### LINKED ISSUE #9820: Refactor Texc to  use JNI

### PR DESCRIPTION
**Pull request by JCash**
This will allow us to use the same JNI generator interface to more easily write bindings between C++ and Java.

Fixes https://github.com/defold/defold/issues/9820

## PR checklist

* [ ] Code
	* [ ] Add engine and/or editor unit tests.
	* [ ] New and changed code follows the overall code style of existing code
	* [ ] Add comments where needed
* [ ] Documentation
	* [ ] Make sure that API documentation is updated in code comments
	* [ ] Make sure that manuals are updated (in github.com/defold/doc)
* [ ] Prepare pull request and affected issue for automatic release notes generator
	* [ ] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [ ] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [ ] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [ ] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [ ] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [ ] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

----------

Example of a well written PR description:

1. Start with the user facing changes. This will end up in the release notes.
1. Add one of the GitHub approved closing keywords
1. Optionally also add the technical changes made. This is information that might help the reviewer. It will not show up in the release notes. Technical changes are identified by a line starting with one of these:
   1. `### Technical changes` 
   1. `Technical changes:`
   2. `Technical notes:`

```
There was a anomaly in the carbon chroniton propeller, introduced in version 8.10.2. This fix will make sure to reset the phaser collector on application startup.

Fixes #1234

### Technical changes
* Pay special attention to line 23 of phaser_collector.clj as it contains some interesting optimizations
* The propeller code was not taking into account a negative phase.
```

### CODE REVIEW

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/TextureGenerator.java [hunk 1]
```diff
@@ -66,44 +67,44 @@ public class TextureGenerator {
...
     private static HashMap<TextureFormatAlternative.CompressionLevel, Integer> compressionLevelLUT = new HashMap<TextureFormatAlternative.CompressionLevel, Integer>();
     static {
-        compressionLevelLUT.put(TextureFormatAlternative.CompressionLevel.FAST, CompressionLevel.CL_FAST);
-        compressionLevelLUT.put(TextureFormatAlternative.CompressionLevel.NORMAL, CompressionLevel.CL_NORMAL);
-        compressionLevelLUT.put(TextureFormatAlternative.CompressionLevel.HIGH, CompressionLevel.CL_HIGH);
-        compressionLevelLUT.put(TextureFormatAlternative.CompressionLevel.BEST, CompressionLevel.CL_BEST);
+        compressionLevelLUT.put(TextureFormatAlternative.CompressionLevel.FAST, Texc.CompressionLevel.CL_FAST.getValue());
+        compressionLevelLUT.put(TextureFormatAlternative.CompressionLevel.NORMAL, Texc.CompressionLevel.CL_NORMAL.getValue());
+        compressionLevelLUT.put(TextureFormatAlternative.CompressionLevel.HIGH, Texc.CompressionLevel.CL_HIGH.getValue());
+        compressionLevelLUT.put(TextureFormatAlternative.CompressionLevel.BEST, Texc.CompressionLevel.CL_BEST.getValue());
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/TextureGenerator.java:73]: Could probably keep the enums until they're actually used, for easier debugging when using toString()

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/TextureGenerator.java [hunk 2]
```diff
@@ -281,8 +288,17 @@ private static TextureImage.Image generateFromColorAndFormat(String name, Buffer
...
-        if (texture == null) {
+        ByteBuffer byteBuffer = getByteBuffer(image);
+        byte[] bytes = byteBuffer.array();
+
+        TimeProfiler.start("CreateTexture");
+        long texture = TexcLibraryJni.CreateTexture(name, width, height,
+                                                        Texc.PixelFormat.PF_A8B8G8R8.getValue(),
+                                                        Texc.ColorSpace.CS_SRGB.getValue(),
+                                                        texcCompressionType, bytes);
+
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/TextureGenerator.java:299]: Some namespace differences, and using byte[] instead of more complex types.

File: engine/graphics/src/transcoder/graphics_transcoder_basisu.cpp [hunk 3]
```diff
@@ -48,7 +48,7 @@ namespace dmGraphics
     }
 
 
-#define TEX_TRANSCODE_DEBUG 1
+#define TEX_TRANSCODE_DEBUG 0
```
Comment (JCash) [inline engine/graphics/src/transcoder/graphics_transcoder_basisu.cpp:51]: Removes the "INFO: Transcoding" messages at runtime

File: engine/jni/src/jni/jni_util.h [hunk 4]
```diff
@@ -136,6 +136,28 @@ namespace dmJNI
...
+        {
+        }
+        ~ScopedByteArray()
+        {
+            if (m_Array && m_JArray)
+            {
+                m_Env->ReleaseByteArrayElements(m_JArray, m_Array, JNI_ABORT);
+            }
+        }
+    };
```
Comment (JCash) [inline engine/jni/src/jni/jni_util.h:159]: Easier management of java byte arrays in C

File: engine/texc/src/java/com/dynamo/bob/pipeline/TexcLibraryJni.java [hunk 5]
```diff
@@ -0,0 +1,488 @@
+//
```
Comment (JCash) [inline engine/texc/src/java/com/dynamo/bob/pipeline/TexcLibraryJni.java:1]: Borrowing much of the same api as TexcLibrary.java

File: engine/texc/src/java/com/dynamo/bob/pipeline/TexcLibraryJni.java [hunk 6]
```diff
@@ -0,0 +1,488 @@
...
+            throw new IllegalArgumentException("Not implemented for data buffer type: " + dataBuffer.getClass());
+        }
+
+        return byteBuffer;
+    }
+
+    // Used for testing the importer. Usage:
+    //    cd engine/modelc
+    //   ./scripts/test_model_importer.sh <model path>
+    public static void main(String[] args) throws IOException, Exception {
```
Comment (JCash) [inline engine/texc/src/java/com/dynamo/bob/pipeline/TexcLibraryJni.java:272]: For easier debugging

File: engine/texc/src/texc_buffer.cpp [hunk 7]
```diff
@@ -59,61 +59,45 @@ namespace dmTexc
         dmZlib::DeflateBuffer(data, size, 9, &out_data_array, DeflateWriter);
         size = out_data_array.Size();
 
-        uint8_t* out_data = (uint8_t*)malloc(size);
-        memcpy(out_data, &out_data_array[0], size);
+        uint8_t* out_data = (uint8_t*)malloc(size + 1);
+        memcpy(out_data+1, &out_data_array[0], size);
```
Comment (JCash) [inline engine/texc/src/texc_buffer.cpp:63]: Previously the encoding was done in a second step (GetBufferData), but now that we only have the CompressBuffer, we do it all in one go, by adding the header byte.

File: editor/src/clj/editor/texture/engine.clj [hunk 8]
```diff
@@ -132,21 +132,22 @@
         (premultiply-alpha texture)                             "could not premultiply alpha"
         (gen-mipmaps texture)                                   "could not generate mip-maps"
         (transcode texture pixel-format SRGB compression-level CT_DEFAULT mipmaps) "could not transcode")
-      (let [buffer-size  (* width-pot height-pot color-count 2)
-            buffer       (little-endian (new-byte-buffer buffer-size))
-            data-size    (TexcLibrary/TEXC_GetData texture buffer buffer-size)
+      (let [;buffer-size  (* width-pot height-pot color-count 2)
+            ;buffer       (little-endian (new-byte-buffer buffer-size))
+            ;data-size    (TexcLibraryJni/GetData texture buffer buffer-size)
+            buffer       (TexcLibraryJni/GetData texture)
```
Comment (Jhonnyg) [inline editor/src/clj/editor/texture/engine.clj:138]: Should this be here?

File: engine/gamesys/src/gamesys/resources/res_font.cpp [hunk 9]
```diff
@@ -152,7 +152,7 @@ namespace dmGameSystem
         uint8_t* data = (uint8_t*)glyph_bank->m_GlyphData.m_Data;
         uint8_t* glyph_data = GetPointer(data, glyph->m_GlyphDataOffset);
 
-        *out_size = glyph->m_GlyphDataSize;
+        *out_size = glyph->m_GlyphDataSize - 1;
```
Comment (Jhonnyg) [inline engine/gamesys/src/gamesys/resources/res_font.cpp:155]: Hm, why is this change needed? Maybe add a comment about it?

File: engine/texc/src/texc.cpp [hunk 10]
```diff
@@ -65,84 +65,99 @@ namespace dmTexc
...
         {
             dmLogWarning("FLIP_AXIS_Z not supported");
             return true;
         }
-        Texture* t = (Texture*) texture;
+        TextureImpl* t = (TextureImpl*)texture->m_Impl;
         return t->m_Encoder.m_FnFlip(t, flip_axis);
     }
 
+    // bool Dither(HTexture texture, PixelFormat pixel_format)
```
Comment (Jhonnyg) [inline engine/texc/src/texc.cpp:149]: junk?

File: engine/texc/src/texclibrary_jni.cpp [hunk 11]
```diff
@@ -0,0 +1,329 @@
...
+
+        JNIFUNC(Encode,                 "(JIIIIZI)Z"),
+
+        // Font glyph buffers
+        JNIFUNC(CompressBuffer,         "([B)L" CLASS_NAME "$Buffer;"),
+
+
+        // Image api
+        //{(char*)"CreateImage", (char*)"(Ljava/lang/String;[BIIII)L" CLASS_NAME "$Scene;", reinterpret_cast<void*>(Java_TexcLibraryJni_CreateImage)},
+    };
```
Comment (Jhonnyg) [inline engine/texc/src/texclibrary_jni.cpp:314]: Is this the "new" api?

**Review by Jhonnyg — DISMISSED** [review]

File: engine/texc/src/texc.cpp (hunk 10 above)
Comment (JCash) [reply engine/texc/src/texc.cpp:149]: Kind of a TODO.
It's part of the next PR (that you've now taken over)

File: engine/texc/src/texclibrary_jni.cpp (hunk 11 above)
Comment (JCash) [reply engine/texc/src/texclibrary_jni.cpp:314]: That was some wip idea for the next PR, will remove from this PR.
See your branch for the new api.

File: engine/gamesys/src/gamesys/resources/res_font.cpp (hunk 9 above)
Comment (JCash) [reply engine/gamesys/src/gamesys/resources/res_font.cpp:155]: It was never accounted for, but the Deflate function handled it anyways.
I'll update the PR
